### PR TITLE
DDF-2094 Add support for decrypting encrypted passwords in sources

### PR DESCRIPTION
**Pull request by roelens8**
#### What does this PR do?
The following sources now support decrypting encrypted passwords:
catalog-opensearch-source
spatial-csw-source
spatial-wfs-v2_0_0-source
spatial-wfs-v1_0_0-source
catalog-registry-api-impl
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@ryeats @coyotesqrl @tbatie @jckilmer @bcwaters @kcwire 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@shaundmorris
@stustison 
#### How should this be tested?
Enter an unencrypted password in a password field of a source configuration. Check the logs and verify it says a clear text password has been passed in. If possible, encrypt the password and pass it in a source configuration and verify the source works.
#### What are the relevant tickets?
DDF-2094
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/OpenSearchSource.java [hunk 1]
```diff
@@ -753,6 +757,9 @@ public String getPassword() {
     }
 
     public void setPassword(String password) {
+        if (encryptionService != null) {
+            encryptionService.decryptValue(password);
```
Comment (coyotesqrl) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/OpenSearchSource.java]: You're not doing anything with the result of the decryption.

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java [hunk 2]
```diff
@@ -92,6 +99,9 @@ public String getPassword() {
     }
 
     public void setPassword(String password) {
+        if (encryptionService != null) {
+            encryptionService.decryptValue(password);
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: Again, ignoring decryption result.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 3]
```diff
@@ -935,6 +940,9 @@ public void setUsername(String username) {
     }
 
     public void setPassword(String password) {
+        if (encryptionService != null) {
+            encryptionService.decryptValue(password);
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Again, ignoring decryption result.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/OpenSearchSource.java (hunk 1 above)
Comment (roelens8) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/ddf/catalog/source/opensearch/OpenSearchSource.java]: I see that now. I fix it in every spot where I decrypt the password.

File: catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v1_0_0/catalog/source/WfsSource.java [hunk 4]
```diff
@@ -280,6 +286,9 @@ public void refresh(Map<String, Object> configuration) throws SecurityServiceExc
 
         if (hasWfsUrlChanged(wfsUrl) || hasDisableCnCheck(disableCnCheckProp)) {
             this.wfsUrl = wfsUrl;
+            if (encryptionService != null) {
+                encryptionService.decryptValue(password);
```
Comment (coyotesqrl) [inline catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v1_0_0/catalog/source/WfsSource.java]: Again, ignoring decryption result.

File: catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java [hunk 5]
```diff
@@ -305,6 +311,9 @@ public void refresh(Map<String, Object> configuration) throws SecurityServiceExc
         super.setId(id);
 
         this.wfsUrl = wfsUrl;
+        if (encryptionService != null) {
+            encryptionService.decryptValue(password);
```
Comment (coyotesqrl) [inline catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java]: Again, ignoring decryption result.

File: catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java [hunk 6]
```diff
@@ -1139,6 +1148,9 @@ public void setUsername(String username) {
     }
 
     public void setPassword(String password) {
+        if (encryptionService != null) {
+            encryptionService.decryptValue(password);
```
Comment (coyotesqrl) [inline catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java]: Again, ignoring decryption result.

**Comment by coyotesqrl** [comment]
You never use the decrypted passwords. Also, should have new tests to run both the with- and without-encryptionService branches, unless it's just an error for there not to be an encryptionService in which case those null tests should all go away.

Also need tests with encrypted passwords, I think.

**Comment by roelens8** [comment]
The encryption service shouldn't ever null, so I will remove all of those null checks.

We should already have units tests for testing the decrypt method using an encrypted password in the EncryptionServiceImplTest. I don't think it would be necessary to create units tests for that in each source as the EncryptionService would have to be mocked anyway.

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java (hunk 2 above)
Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: Fixed.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java (hunk 3 above)
Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Fixed.

File: catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v1_0_0/catalog/source/WfsSource.java (hunk 4 above)
Comment (roelens8) [inline catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v1_0_0/catalog/source/WfsSource.java]: Fixed.

File: catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java (hunk 5 above)
Comment (roelens8) [inline catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java]: Fixed.

File: catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java (hunk 6 above)
Comment (roelens8) [inline catalog/spatial/wfs/2.0.0/spatial-wfs-v2_0_0-source/src/main/java/org/codice/ddf/spatial/ogc/wfs/v2_0_0/catalog/source/WfsSource.java]: Fixed.

**Comment by shaundmorris** [comment]
I had a minor concern that the decrypt() method might return null when getting an empty string passed in. However, looks like sources adequately handle both cases. If not I was going to suggest adding tests there. Not sure if that is what you and @coyotesqrl were referring to (since I'm looking at the PR after changes) on null checks

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java [hunk 7]
```diff
@@ -92,7 +103,11 @@ public String getPassword() {
     }
 
     public void setPassword(String password) {
-        this.password = password;
+        String updatedPassword = password;
+        if (encryptionService != null && !password.equals("")) {
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: Do you still need this test on `encryptionService != null`?

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java [hunk 8]
```diff
@@ -92,7 +103,11 @@ public String getPassword() {
     }
 
     public void setPassword(String password) {
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: Couldn't this whole method be crystallized down to...
```java
public void setPassword(String password) {
    if (!StringUtils.isEmpty(password)) {
        this.password = encryptionService.decryptValue(password);
    }
}
```
This also protects you from a null password passed in on an update.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 9]
```diff
@@ -935,7 +951,11 @@ public void setUsername(String username) {
     }
 
     public void setPassword(String password) {
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: My same comment on simplifying this method.

**Comment by coyotesqrl** [comment]
Simplification/cleanup comments aside, 👍

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java (hunk 7 above)
Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: I still need a null check for this method in CswSourceConfiguration and the setPassword method in AbstractCswSource. This is because as I am deprecating the default constructors in those classes, so there is a possibility the encryption service could be null if there is some code calling those constructors.

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java (hunk 8 above)
Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: Added the StringUtils.isNotEmpty(password) check along with keeping the null check of the encryption service.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java (hunk 9 above)
Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: My same comment here also.

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java (hunk 8 above)
Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: The example you mentioned prevents a user from updating the password with an empty string.We should still allow a user to update it to empty string.

The reason for checking if the string is not empty in the first place is because the encryption service returns null if an empty string is passed in.

Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: Better yet, I'll update the EncryptionService to return an empty string if an empty string is passed in the decrypt method.

Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: I'm not sure that's the right approach. @stustison do you think the service should be changed in that way?

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/TestCswSource.java [hunk 10]
```diff
@@ -945,26 +960,78 @@ public void testRefresh() throws SecurityServiceException {
...
 
         // Set Configuration Map
         Map<String, Object> configuration = getConfigurationMap(cswSource);
+        when(encryptionService.decryptValue(PASSWORD)).thenReturn(PASSWORD);
 
         // Call Refresh
         cswSource.refresh(configuration);
 
-        // Get Configuration
+        // Get Configuration"
```
Comment (shaundmorris) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/TestCswSource.java]: @roelens8 think you have an extra double quote in the comment here

Comment (roelens8) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/TestCswSource.java]: Fixed

File: catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java (hunk 8 above)
Comment (stustison) [inline catalog/spatial/csw/spatial-csw-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/CswSourceConfiguration.java]: I'm not a fan of passing back nulls from any sort of API with some caveats. In my opinion, we should pass back from the method whatever it receives. If we pass in an unencrypted password, we should pass back that password. Empty string is an unencrypted password and we should just pass that back as well. Same also goes for null. If null is passed in, we should probably just hand null right back.

**Comment by kcwire** [comment]
👍

**Comment by ryeats** [comment]
hero'd 👍
